Add paper for JuliaCon proceedings

ayushpatnaikgit opened the pull request:
I have dumped most of the stuff from the slideshow. 

@smishr could you take it from here? 

This is the author's guide: https://juliacon.github.io/proceedings-guide/author/

The limit is 2 pages.

ayushpatnaikgit committed:
Adding paper branch and associated files.

ayushpatnaikgit committed:
Added most of the content from the JuliaCon2023 presentation

ayushpatnaikgit committed:
Finished adding all the content from the slideshow. Added related works.

ayushpatnaikgit committed:
Added equation number to make it consistent. All the equation numbers could be removed.

smishr commented:
Ill have a look tomoroww

ayushpatnaikgit committed:
"and" should be fixed in the equation.

ayushpatnaikgit committed:
Prepare the first draft of the paper. Few lines need to be cut or pushed to the footnotes.

ayushpatnaikgit committed:
Fit in 2 pages.

ayushpatnaikgit committed:
Remove subsection for example

ayushpatnaikgit committed:
First draft

ayushpatnaikgit committed:
Minor changes

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+\begin{abstract}
+
+Estimating variances in survey data analysis is challenging due to the complex nature of survey designs. It is typically done through resampling methods like bootstrapping. The Survey.jl package leverages Julia to provide an efficient framework for these resampling techniques, facilitating faster survey data analysis.
```
```suggestion
Estimating variances in survey data analysis is challenging due to the complex nature of survey designs. It is typically done through resampling methods like bootstrapping, which can be computationally intensive. The Survey.jl package leverages Julia to provide an efficient framework for these resampling techniques, facilitating faster survey data analysis.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+\section{Introduction}
+
+The growing volume of survey datasets necessitate more efficient analysis methods, particularly for variance estimation in complex survey designs. Computationally demanding resampling techniques, like bootstrapping and jackknife, are required when dealing with stratification, clustering, and unequal weights. 
```
```suggestion
The growing volume of survey datasets necessitates more efficient analysis methods, particularly for variance estimation in complex survey designs. Computationally demanding resampling techniques, such as bootstrapping and jackknife, are required when dealing with stratification, clustering, and unequal weights. 
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+The growing volume of survey datasets necessitate more efficient analysis methods, particularly for variance estimation in complex survey designs. Computationally demanding resampling techniques, like bootstrapping and jackknife, are required when dealing with stratification, clustering, and unequal weights. 
+\\
+
+Many software packages exist for survey analysis\footnote{A comprehensive list is provided by \cite{SummarySurveyAnalysis}}. Notable examples include the R survey package, SAS/STAT, SPSS Complex Samples, Stata, and SUDAAN. The R survey package by Thomas Lumley\cite{lumley2004analysis} is widely recognized for its comprehensive capabilities and open-source availability. However, it lacks computati ...
```
```suggestion
Many software packages exist for survey analysis\footnote{A comprehensive list is provided by \cite{SummarySurveyAnalysis}}. Notable examples include the R survey package, SAS/STAT, SPSS Complex Samples, Stata, and SUDAAN. The R survey package by Thomas Lumley\cite{lumley2004analysis} is widely recognized for its comprehensive capabilities and open-source availability. However, it is limited by R's computational efficiency, especially for large-scale data. Survey.jl leverages Julia to offer a faster resampling framework for variance estimation and survey data analysis.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+\section{Survey design}
+
+A \verb|SurveyDesign| object can be created to incorporate the sampling design. This object requires the following parameters: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the ...
```
```suggestion
The \verb|SurveyDesign| object instantiates a sampling design. This object requires the following parameters: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the column name containing the sampling weights; and \verb|popsize::Symbol|, indicating the column name containing the population size.\footnote{Internaly, there is a single constructor for all types of surveys. Every survey is assumed to be a complex survey. If there is no stratification, we assume that everything is part of one stratum. If there is no clustering, we assume each member is a cluster.}  
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+Survey.jl provides basic univariate and multivariate estimators for efficient survey data analysis.
+
+For univariate statistics such as mean, median, total, and quantiles, the following examples illustrate their usage:
```
```suggestion
Survey.jl provides a range of estimators for survey data analysis. These include univariate statistics such as mean, median, total, and quantiles, as well as multivariate statistics such as regressions and ratios. For example, to estimate the mean of the \verb|:api99| column in the \verb|design| SurveyDesign:
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+    \end{lstlisting}
+This command estimates the mean of column \verb|:api99|.
+
+For multivariate statistics such as regressions\footnote{Regressions are performed using GLM.jl. Instead of passing a DataFrame, a survey design is passed to the function, maintaining a familiar interface. This approach of using multiple dispatch is applied to all estimators imported from other packages, ensuring consistency and ease of use.}:
```
```suggestion
Regression models\footnote{Regressions are performed using GLM.jl. Instead of passing a DataFrame, a survey design is passed to the function, maintaining a familiar interface. This approach of using multiple dispatch is applied to all estimators imported from other packages, ensuring consistency and ease of use.} can be fitted directly to \verb|SurveyDesign| objects:
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+For univariate statistics such as mean, median, total, and quantiles, the following examples illustrate their usage:
+
+\begin{lstlisting}
+julia> mean(:api99, design)
```
```suggestion
julia> mean(:api99, my_design)
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+For multivariate statistics such as regressions\footnote{Regressions are performed using GLM.jl. Instead of passing a DataFrame, a survey design is passed to the function, maintaining a familiar interface. This approach of using multiple dispatch is applied to all estimators imported from other packages, ensuring consistency and ease of use.}:
+
+\begin{lstlisting}
+julia> glm(@formula(y ~ x),
```
```suggestion
julia> glm(@formula(:api99 ~ :api00),
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+\section{Replicate weights}
+
+The standard error of an estimator measures the average amount of variability or uncertainty in the estimated value. Standard errors are often provided alongside point estimates in various statistical packages.
+
+To estimate standard errors for complex survey designs, Survey.jl uses replicate weights, generated through resampling techniques such as bootstrap and jackknife. Each replicate sample represents a plausible variation of the original sample, allowing for the estimation of variability as if the sampling were repeated multiple times.
```
```suggestion
Survey data analysis requires variance estimation to quantify the uncertainty in the estimates. This is particularly important for complex survey designs, where the standard error of an estimator is not directly available. 

Survey.jl provides a framework for estimating standard errors by generating replicate weights through resampling techniques such as bootstrapping and jackknife. Each replicate sample represents a plausible variation of the original sample, allowing for the estimation of variability as if the sampling were repeated multiple times.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+To estimate standard errors for complex survey designs, Survey.jl uses replicate weights, generated through resampling techniques such as bootstrap and jackknife. Each replicate sample represents a plausible variation of the original sample, allowing for the estimation of variability as if the sampling were repeated multiple times.
+
+The estimate is calculated for each replicate, and then the standard error is computed from the distribution of these estimates. 
```
```suggestion
An estimate of the variance is obtained by calculating the variance of the estimates across the replicates.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+Where \( w_i(r) \) denotes the initial weight for observation \( i \) within replicate \( r \), \( n_h \) is the total number of observations in stratum \( h \), and \( m_{h}(r) \) is the number of PSUs in stratum \( h \) that are selected in replicate \( r \)\cite{Lohr}.
+
+\verb|bootweights| can be used to generate \verb|ReplicateDesign{BootstrapReplicates}| from a \verb|SurveyDesign|. 
```
```suggestion
Survey.jl supports bootstrapping for variance estimation using the \verb|bootweights| function. This function generates replicate weights for bootstrapping from a \verb|SurveyDesign| object, and the resulting object is a \verb|ReplicateDesign{BootstrapReplicates}|.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+\verb|bootweights| can be used to generate \verb|ReplicateDesign{BootstrapReplicates}| from a \verb|SurveyDesign|. 
+
+\begin{lstlisting}
+julia> bdesign = bootweights(design; replicates = 1000)
```
```suggestion
julia> my_bootstrapped_design = bootweights(my_design; replicates = 1000)
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+% replicates: 1000
+% \end{lstlisting}
+
+The replicate design object facilitates variance estimation. When a function receives a \verb|ReplicateDesign| rather than a \verb|SurveyDesign|, it provides the standard error along with the point estimate.
```
```suggestion
\verb|ReplicateDesign| objects can be passed to summary statistics functions, which will return the standard error of the estimator alongside the point estimate:
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+The replicate design object facilitates variance estimation. When a function receives a \verb|ReplicateDesign| rather than a \verb|SurveyDesign|, it provides the standard error along with the point estimate.
+For example: 
+\begin{lstlisting}
+julia> mean(:api99, bdesign)
```
```suggestion
julia> mean(:api99, my_bootstrapped_design)
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+--------|-----------------
+      1 | 606.978  24.7505
+    \end{lstlisting}
+For each replicate $r$, $\hat{\theta}^*_r$ is the estimator of $\theta$, calculated the same way as $\hat{\theta}$ but using weights $w_i'(r)$ instead of the original weights $w_i$. The variance of the estimator is given by: 
```
```suggestion
For each replicate $r$, $\hat{\theta}^*_r$ is calculated using the adjusted weights $w_i'(r)$. $\hat{\theta}^*$ instead of the original weights $w_i$. The variance of the estimator is estimated as: 
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+
+\subsection{Jackknife}
+In the jackknife method, each PSU is systematically omitted one at a time to create replicates. The adjusted weight \( w_{i(hj)} \) for observation \( i \) when PSU \( j \) in stratum \( h \) is omitted is calculated as:
```
```suggestion
Survey.jl also supports the jackknife method for variance estimation. The package implements the leave-one-out jackknife, which omits each PSU one at a time to create replicates. The adjusted weight \( w_{i(hj)} \) for observation \( i \) when PSU \( j \) in stratum \( h \) is omitted is calculated as:
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+    \end{cases} %% Fix equation
+    \end{equation} \cite{Lohr}
+
+\verb|jackknifeweights| can be used to generate \verb|ReplicateDesign{JackknifeReplicates}| from a \verb|SurveyDesign|. 
```
```suggestion
The \verb|jackknifeweights| function takes in a \verb|SurveyDesign| object and returns a \verb|ReplicateDesign{JackknifeReplicates}| object:
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+\verb|jackknifeweights| can be used to generate \verb|ReplicateDesign{JackknifeReplicates}| from a \verb|SurveyDesign|. 
+
+\begin{lstlisting}
+    julia> design = jackknifeweights(design)
```
```suggestion
    julia> my_jackknife_design = jackknifeweights(my_design)
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+% replicates: 200
+% \end{lstlisting}
+
+This object can be passed to estimators to obtain an estimate of variance alongside the point estimate. 
```
```suggestion
This \verb|ReplicateDesign| object can also be passed to summary statistics functions to obtain an estimate of the variance of the estimator alongside its point estimate.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+\subsection{Extending variance estimation}
+
+Survey.jl supports variance estimation for basic estimators such as mean, quantile, ratio, and regressions. The package is designed to extend this capability to any estimator through a generalized approach.
+
+The \verb|Survey.variance| function can be applied to replicate designs to estimate the variance of any estimator function, allowing users and developers to extend variance estimation to custom estimators.
```
```suggestion
Survey.jl currently supports variance estimation for the summary statistics functions provided by the package, but the framework can be extended to custom estimators. The \verb|variance| function can be applied to \verb|ReplicateDesign| objects to estimate the variance of an estimator function, such as \verb|Survey.mean|.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+% at appropriate place in your \TeX{} file or in bibliography file.
+
+\section{Conclusions}
+Survey.jl offers an efficient framework for survey data analysis. Its functionality has been tested against R's survey package, and future development aims to port all features from R.
```
```suggestion
Survey.jl provides a comprehensive framework for survey data analysis, leveraging Julia's computational efficiency. The package has been tested against R's survey package, and future development aims to port all features from R.
```

nadiaenh commented on paper/paper.tex:
```diff
@@ -0,0 +1,272 @@
+
+This object can be passed to estimators to obtain an estimate of variance alongside the point estimate. 
+
+$\hat{\theta}$ represents the estimator computed using the original weights, and $\hat{\theta_{(hj)}}$ represents the estimator computed from the replicate weights obtained when PSU $j$ from cluster $h$ is removed. The variance is given by: 
```
```suggestion
$\hat{\theta}$ represents the estimator computed using the original weights, and $\hat{\theta_{(hj)}}$ represents the estimator computed from the replicate weights obtained when PSU $j$ from cluster $h$ is removed. The variance is estimated as: 
```

nadiaenh reviewed (commented):
Also I think you should remove the API example, and instead add the output of the `glm` function call.

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit committed:
Update paper/paper.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

ayushpatnaikgit commented:
> Also I think you should remove the API example, and instead add the output of the `glm` function call.

Thanks @nadiaenh for the comments! 

I wanted to use the NHANES example, but it doesn't have a numerical variable column.

nadiaenh committed:
Update header.tex

Added co-author

nadiaenh commented on paper/header.tex:
```diff
@@ -0,0 +1,19 @@
+
+\hypersetup{
+pdftitle = {Survey.jl - An Efficient Framework for Analysing Complex Surveys},
+pdfsubject = {JuliaCon 2022 Proceedings},
```
```suggestion
pdfsubject = {JuliaCon 2023 Proceedings},
```

nadiaenh committed:
Add co-author to paper.yml

JuliaCon Proceedings Author's Guide says we should only edit these 3 files because other files might be overwritten later:
`paper.tex`
`ref.bib`
`paper.yml`

ayushpatnaikgit committed:
Update paper/header.tex

Co-authored-by: Nadia <76887318+nadiaenh@users.noreply.github.com>

codetalker7 commented on paper/paper.tex:
```diff
@@ -0,0 +1,268 @@
+
+\section{Survey design}
+
+A \verb|SurveyDesign| object can be created to incorporate the sampling design. This object requires the following parameters: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the ...
```
`This object requires the following parameters`: instead of `parameters`, should we say `properties` instead? Usually parameters are constants/values that tag some method/type, or I like to think of them as tunable values associated to some model.

ayushpatnaikgit replied:
How about `argument`?

codetalker7 replied:
`argument` definitely sounds better.

ayushpatnaikgit replied:
> How about `argument`?

```suggestion
A \verb|SurveyDesign| object can be created to incorporate the sampling design. This object requires the following arguments: \verb|data::DataFrame|, which is the survey data in the form of a \verb|DataFrame|; \verb|clusters::Symbol|, specifying the column name containing the clusters; \verb|strata::Symbol|, specifying the column name containing the strata; \verb|weights::Symbol|, indicating the column name containing the sampling weights; and \verb|popsize::Symbol|, indicating the column name containing the population size.\footnote{Internaly, there is a single constructor for all types of surveys. Every survey is assumed to be a complex survey. If there is no stratification, we assume that everything is part of one stratum. If there is no clustering, we assume each member is a cluster.}  
```

codetalker7 commented on paper/paper.tex:
```diff
@@ -0,0 +1,268 @@
+julia> apiclus1 = load_data("apiclus1"); 
+# CSV dataframe included with the package
+
+design = SurveyDesign(apiclus1; 
```
`julia> ` missing here?

ayushpatnaikgit replied:
> `julia> ` missing here?

```suggestion
julia> design = SurveyDesign(apiclus1); 
```

codetalker7 commented on paper/paper.tex:
```diff
@@ -0,0 +1,268 @@
+
+\begin{lstlisting}
+julia> glm(@formula(y ~ x),
+            my_design, Normal(), IdentityLink())
```
A semicolon missing? Since we're not showing any output?

ayushpatnaikgit replied:
> A semicolon missing? Since we're not showing any output?

```suggestion
            my_design, Normal(), IdentityLink()); 
```

codetalker7 commented on paper/paper.tex:
```diff
@@ -0,0 +1,268 @@
+    w_i'(r) = w_i(r) \times \frac{n_h}{n_h - 1} \times m_{h}(r)
+\end{equation}
+
+Where \( w_i(r) \) denotes the initial weight for observation \( i \) within replicate \( r \), \( n_h \) is the total number of observations in stratum \( h \), and \( m_{h}(r) \) is the number of PSUs in stratum \( h \) that are selected in replicate \( r \)\cite{Lohr}.
```
Some space between `\( r \)` and `\cite{Lohr}` will be nice! Initially I thought we're indexing the variable `r`!

ayushpatnaikgit replied:
> Some space between `\( r \)` and `\cite{Lohr}` will be nice! Initially I thought we're indexing the variable `r`!

```suggestion
Where \( w_i(r) \) denotes the initial weight for observation \( i \) within replicate \( r \), \( n_h \) is the total number of observations in stratum \( h \), and \( m_{h}(r) \) is the number of PSUs in stratum \( h \) that are selected in replicate \( r \) \cite{Lohr}.
```

codetalker7 commented on paper/paper.yml:
```diff
@@ -0,0 +1,21 @@
+    index: 1
+  - name: Simon Fraser University
+    index: 2
+date: 24 April 2019
```
Is this some typo (the year is 2019)?

ayushpatnaikgit replied:
> Is this some typo (the year is 2019)?



> Is this some typo (the year is 2019)?

```suggestion
date: 27 May 2024
```

codetalker7 committed:
Adding myself as co-author.

ayushpatnaikgit commented:
@smishr @iuliadmtru 

Please add yourself as co-authors if you are interest. Directly commit to the `yml` file.

ayushpatnaikgit committed:
Update paper/paper.yml

ayushpatnaikgit committed:
Update paper/paper.tex

ayushpatnaikgit committed:
Update paper/paper.tex

ayushpatnaikgit committed:
Update paper/paper.tex

ayushpatnaikgit committed:
Update paper/paper.tex

ayushpatnaikgit commented on paper/paper.yml:
```diff
@@ -0,0 +1,25 @@
+    affiliation: 2
+  - name: Siddhant Chaudhary
+    affiliation: 3
+  # - name: 4th author
+  #   affiliation: 4
```
```suggestion
   - name: Shikhar Mishra
     affiliation: 1
```

ayushpatnaikgit replied:
@smishr just click on commit to give your consent.

smishr committed:
Update paper/paper.yml

added Shikhar

Co-authored-by: Ayush Patnaik <ayushpatnaik@gmail.com>

smishr commented:
done

ayushpatnaikgit committed:
Fix formatting

ayushpatnaikgit committed:
Remove comments

ayushpatnaikgit commented:
Thanks everyone, submitting the paper. Will ask the editors to allow Iulia to add her name in the subsequent stages.

ayushpatnaikgit commented:
@iuliadmtru right now, your name is not on the paper, but I have requested the editor to allow us to add it in the future.
